[JSC] Add async stack traces behind the flag

sosukesuzuki opened the pull request:
#### 74c38785fa38f8debe3dcec803b13bd44843b242
<pre>
[JSC] Add async stack traces behind the flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=296261">https://bugs.webkit.org/show_bug.cgi?id=296261</a>

Reviewed by Yusuke Suzuki.

This reverts <a href="https://commits.webkit.org/299498@main">https://commits.webkit.org/299498@main</a> to reland <a href="https://commits.webkit.org/299482@main">https://commits.webkit.org/299482@main</a>

------

JSC cannot obtain sufficient async stack traces. For example, this code:

    async function one(x) {
        await two(x);
    }
    async function two(x) {
        await x;
        throw new Error(&quot;error from two&quot;);
    }
    one(1).catch((err) =&gt; { print(err.stack); });

prints this stack trace:

    two@./test.js:6:20

When an error occurs within an async function, the stack trace is cut off
at the most recent await point, so higher-level functions in the async call
chain are not displayed. This happens because when an async function is
suspended at an await statement and later resumed from the microtask queue,
the original call stack is lost.

While this behavior is correct from the semantics of call stacks and microtask
queues, it is inconvenient for users.

This patch builds async stack traces through the following approach (only when
the --useAsyncStackTrace flag is enabled):

- Detect calls to `@promiseReactionJob` and `@promiseReactionJobWithoutPromise`
  during stack trace construction and retrieve the JSGenerator object hidden
  behind async-await from its arguments
- Obtain references to awaited Promises held in each generator&apos;s Context field
- Trace async context (parent generators recorded as @context fields) from the
  Promise&apos;s reaction chain
- Walk up the chain of parent generators and construct stack frames from function
  information stored in each async function&apos;s Next field

After this patch, the previous code outputs this stack trace:

    two@./WebKitBuild/Debug/test.js:6:20
    one@./WebKitBuild/Debug/test.js:2:14

V8 constructs async stack traces using a similar approach [1][2].

[1]: <a href="https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit?tab=t.0#heading=h.9ss45aibqpw2">https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit?tab=t.0#heading=h.9ss45aibqpw2</a>
[2]: <a href="https://issues.chromium.org/issues/42210758">https://issues.chromium.org/issues/42210758</a>

* JSTests/stress/async-stack-trace-basic.js: Added.
(nop):
(shouldThrowAsync):
(throw.new.Error.async one):
(throw.new.Error.async two):
(async one):
(async two):
(throw.new.Error.async drainMicrotasks):
(callOne):
(callTwo):
(async drainMicrotasks):
(async throwError):
* JSTests/stress/async-stack-trace-bound-functions-basic.js: Added.
(nop):
(shouldThrowAsync):
(throw.new.Error.async foo):
(throw.new.Error.async bar):
(throw.new.Error.async baz):
* JSTests/stress/async-stack-trace-nested-deep.js: Added.
(nop):
(shouldThrowAsync):
(async level1):
(async level2):
(async level3):
(async level4):
(async level5):
(async level6):
(async level7):
(async level8):
(async level9):
(async level10):
(async level11):
(async level12):
(async level13):
(async level14):
(async level15):
(async level16):
(async level17):
(async level18):
(async level19):
(async level20):
(async level21):
(async level22):
(async level23):
(async level24):
(async level25):
(async level26):
(async level27):
(async level28):
(async level29):
(async level30):
(async level31):
(async level32):
(async level33):
(async level34):
(async level35):
(async level36):
(async level37):
(async level38):
(async level39):
(async level40):
(async level41):
(async level42):
(async level43):
(async level44):
(async level45):
(async level46):
(async level47):
(async level48):
(async level49):
(async level50):
(async level51):
(async level52):
(async level53):
(async level54):
(async level55):
(async level56):
(async level57):
(async level58):
(async level59):
(async level60):
(async level61):
(async level62):
(async level63):
(async level64):
(async level65):
(async level66):
(async level67):
(async level68):
(async level69):
(async level70):
(async level71):
(async level72):
(async level73):
(async level74):
(async level75):
(async level76):
(async level77):
(async level78):
(async level79):
(async level80):
(async level81):
(async level82):
(async level83):
(async level84):
(async level85):
(async level86):
(async level87):
(async level88):
(async level89):
(async level90):
(async level91):
(async level92):
(async level93):
(async level94):
(async level95):
(async level96):
(async level97):
(async level98):
(async level99):
(async level100):
(async level101):
(async delayedOperation):
(async problematicFunction):
* JSTests/stress/async-stack-trace-nested.js: Added.
(nop):
(shouldThrowAsync):
(async level1):
(async level2):
(async level3):
(async level4):
(async level5):
(async level6):
(async level7):
(async level8):
(async level9):
(async level10):
(async delayedOperation):
(async problematicFunction):
* JSTests/stress/async-stack-trace-switch.js: Added.
(shouldThrowAsync):
(throw.new.Error.async foo):
(throw.new.Error.async bar):
(throw.new.Error.async baz):
* JSTests/stress/never-inlined-functions.js: Added.
(shouldBe):
(const.regularFunction):
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.neverInline.promiseReactionJob):
(linkTimeConstant.promiseReactionJobWithoutPromise): Deleted.
(linkTimeConstant.promiseReactionJob): Deleted.
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
(JSC::Interpreter::getStackTrace):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseReactionJobFunction const):
(JSC::JSGlobalObject::promiseReactionJobWithoutPromiseFunction const):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/StackFrame.cpp:
(JSC::StackFrame::StackFrame):
(JSC::StackFrame::sourceURL const):
(JSC::StackFrame::sourceURLStripped const):
* Source/JavaScriptCore/runtime/StackFrame.h:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):

Canonical link: <a href="https://commits.webkit.org/299595@main">https://commits.webkit.org/299595@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c3f94b1c8f145e5a7ceb8eb5dd6a2cad77a5f60

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119381 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39068 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29723 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125629 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71452 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121258 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47649 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90711 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60029 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122333 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31731 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107036 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71169 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30766 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25145 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69282 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111504 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101185 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128626 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117897 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46299 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35037 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99288 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46664 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103235 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99082 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25205 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44525 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22530 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42866 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46162 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51862 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146597 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45627 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37692 "Found 1 new JSC binary failure: testapi, Found 18677 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/StackArgsWithFormals.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48977 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/builtins/PromiseOperations.js:
```diff
@@ -452,7 +454,15 @@ function promiseReactionJob(promiseOrCapability, handler, argument, contextOrSta
     // Case (4).
     if (!promiseOrCapability) {
-        @promiseReactionJobWithoutPromise(handler, argument, contextOrState);
+        // inlined promiseReactionJobWithPromise
+        try {
+            if (contextOrState)
+                handler(argument, contextOrState);
+            else
+                handler(argument);
+        } catch {
+            // This is user-uncatchable promise. We just ignore the error here.
+        }
```
Based on the discussion on Slack (https://webkit.slack.com/archives/CTV4FGWF4/p1756915196720469), I'm manually inlining the call to `@promiseReactionJobWithoutPromise` by `@promiseReactionJob`. When I ran JetStream/doxbee-async locally, this indeed improves the score.

Constellation reviewed (approved):
r=me if perf is recovered

sosukesuzuki committed:
[JSC] Add async stack traces behind the flag
https://bugs.webkit.org/show_bug.cgi?id=296261

Reviewed by Yusuke Suzuki.

This reverts https://commits.webkit.org/299498@main to reland https://commits.webkit.org/299482@main

------

JSC cannot obtain sufficient async stack traces. For example, this code:

    async function one(x) {
        await two(x);
    }
    async function two(x) {
        await x;
        throw new Error("error from two");
    }
    one(1).catch((err) => { print(err.stack); });

prints this stack trace:

    two@./test.js:6:20

When an error occurs within an async function, the stack trace is cut off
at the most recent await point, so higher-level functions in the async call
chain are not displayed. This happens because when an async function is
suspended at an await statement and later resumed from the microtask queue,
the original call stack is lost.

While this behavior is correct from the semantics of call stacks and microtask
queues, it is inconvenient for users.

This patch builds async stack traces through the following approach (only when
the --useAsyncStackTrace flag is enabled):

- Detect calls to `@promiseReactionJob` and `@promiseReactionJobWithoutPromise`
  during stack trace construction and retrieve the JSGenerator object hidden
  behind async-await from its arguments
- Obtain references to awaited Promises held in each generator's Context field
- Trace async context (parent generators recorded as @context fields) from the
  Promise's reaction chain
- Walk up the chain of parent generators and construct stack frames from function
  information stored in each async function's Next field

After this patch, the previous code outputs this stack trace:

    two@./WebKitBuild/Debug/test.js:6:20
    one@./WebKitBuild/Debug/test.js:2:14

V8 constructs async stack traces using a similar approach [1][2].

[1]: https://docs.google.com/document/d/13Sy_kBIJGP0XT34V1CV3nkWya4TwYx9L3Yv45LdGB6Q/edit?tab=t.0#heading=h.9ss45aibqpw2
[2]: https://issues.chromium.org/issues/42210758

* JSTests/stress/async-stack-trace-basic.js: Added.
(nop):
(shouldThrowAsync):
(throw.new.Error.async one):
(throw.new.Error.async two):
(async one):
(async two):
(throw.new.Error.async drainMicrotasks):
(callOne):
(callTwo):
(async drainMicrotasks):
(async throwError):
* JSTests/stress/async-stack-trace-bound-functions-basic.js: Added.
(nop):
(shouldThrowAsync):
(throw.new.Error.async foo):
(throw.new.Error.async bar):
(throw.new.Error.async baz):
* JSTests/stress/async-stack-trace-nested-deep.js: Added.
(nop):
(shouldThrowAsync):
(async level1):
(async level2):
(async level3):
(async level4):
(async level5):
(async level6):
(async level7):
(async level8):
(async level9):
(async level10):
(async level11):
(async level12):
(async level13):
(async level14):
(async level15):
(async level16):
(async level17):
(async level18):
(async level19):
(async level20):
(async level21):
(async level22):
(async level23):
(async level24):
(async level25):
(async level26):
(async level27):
(async level28):
(async level29):
(async level30):
(async level31):
(async level32):
(async level33):
(async level34):
(async level35):
(async level36):
(async level37):
(async level38):
(async level39):
(async level40):
(async level41):
(async level42):
(async level43):
(async level44):
(async level45):
(async level46):
(async level47):
(async level48):
(async level49):
(async level50):
(async level51):
(async level52):
(async level53):
(async level54):
(async level55):
(async level56):
(async level57):
(async level58):
(async level59):
(async level60):
(async level61):
(async level62):
(async level63):
(async level64):
(async level65):
(async level66):
(async level67):
(async level68):
(async level69):
(async level70):
(async level71):
(async level72):
(async level73):
(async level74):
(async level75):
(async level76):
(async level77):
(async level78):
(async level79):
(async level80):
(async level81):
(async level82):
(async level83):
(async level84):
(async level85):
(async level86):
(async level87):
(async level88):
(async level89):
(async level90):
(async level91):
(async level92):
(async level93):
(async level94):
(async level95):
(async level96):
(async level97):
(async level98):
(async level99):
(async level100):
(async level101):
(async delayedOperation):
(async problematicFunction):
* JSTests/stress/async-stack-trace-nested.js: Added.
(nop):
(shouldThrowAsync):
(async level1):
(async level2):
(async level3):
(async level4):
(async level5):
(async level6):
(async level7):
(async level8):
(async level9):
(async level10):
(async delayedOperation):
(async problematicFunction):
* JSTests/stress/async-stack-trace-switch.js: Added.
(shouldThrowAsync):
(throw.new.Error.async foo):
(throw.new.Error.async bar):
(throw.new.Error.async baz):
* JSTests/stress/never-inlined-functions.js: Added.
(shouldBe):
(const.regularFunction):
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(linkTimeConstant.neverInline.promiseReactionJob):
(linkTimeConstant.promiseReactionJobWithoutPromise): Deleted.
(linkTimeConstant.promiseReactionJob): Deleted.
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::getAsyncStackTrace):
(JSC::Interpreter::getStackTrace):
* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::promiseReactionJobFunction const):
(JSC::JSGlobalObject::promiseReactionJobWithoutPromiseFunction const):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/StackFrame.cpp:
(JSC::StackFrame::StackFrame):
(JSC::StackFrame::sourceURL const):
(JSC::StackFrame::sourceURLStripped const):
* Source/JavaScriptCore/runtime/StackFrame.h:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):

Canonical link: https://commits.webkit.org/299595@main
